src/modules/legacyBalancesProvider.js

```diff
@@ -0,0 +1,51 @@
+import store from "modules/store";
+import { setupWatch } from "./web3Provider";
+import { refreshLegacyBalances } from "modules/reducers/legacyBalances";
+
+export default () => {
+    const web3Connect = store.getState().web3Connect;
+
+    if (web3Connect.isConnected) {
+        console.debug("legacyBalancesProvider - web3 already connected, dispatching refreshLegacyBalances() ");
+        store.dispatch(refreshLegacyBalances());
+    }
+
+    const monetarySupervisor = store.getState().monetarySupervisor;
+
+    if (monetarySupervisor.isConnected) {
+        setupListeners();
+    }
+
+    setupWatch("monetarySupervisor.contract", onMonetarySupervisorContractChange);
+    setupWatch("web3Connect.userAccount", onUserAccountChange);
+
+    return;
+};
+
+const onMonetarySupervisorContractChange = (newVal, oldVal) => {
+    if (newVal && store.getState().augmintToken.isConnected) {
+        setupListeners();
+    }
+};
+
+const setupListeners = () => {
+    const monetarySupervisor = store.getState().monetarySupervisor.contract.ethersInstance;
+    monetarySupervisor.onlegacytokenconverted = onLegacyTokenConverted;
+};
+
+const onLegacyTokenConverted = (oldTokenAddress, account, amount) => {
+    const userAccount = store.getState().web3Connect.userAccount;
+    if (account.toLowerCase() === userAccount.toLowerCase()) {
+        console.debug(
+            "legacyBalancesProvider - LegacyTokenConverted event received for current user account. Dispatching refreshLegacyBalances()"
+        );
+        store.dispatch(refreshLegacyBalances());
+    }
+};
+
+const onUserAccountChange = (newVal, oldVal, objectPath) => {
+    if (newVal !== "?") {
```

Comment:
is this condition for the initial account change?
would `(newVal !== oldVal)` make sense here?

Reply:
while loading it has ? value (see reducers/webConnect.js)

Reply:
yeah, but there's no case when it changes after?

Reply:
TBH I don't remember why is it like that. maybe until intialised it's undefined or null then after initialisation it set to "?" which triggers change event?
 We should revisit all providers , it became convoluted and newer metamask / web3 versions might have made some parts unecessery